src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+
+@[simp] lemma coe_mk0 {a : G} (h : a ≠ 0) : (mk0 a h : G) = a := rfl
+
+@[simp] theorem inv_eq_inv (u : units G) : (↑u⁻¹ : G) = u⁻¹ :=
+(mul_left_inj u).1 $ by { rw [units.mul_inv, gwz.mul_inv_cancel], apply gwz.unit_ne_zero }
+
+@[simp] lemma mk0_coe (u : units G) (h : (u : G) ≠ 0) : mk0 (u : G) h = u :=
+units.ext rfl
+
+@[simp] lemma mk0_inj {a b : G} (ha : a ≠ 0) (hb : b ≠ 0) :
+  units.mk0 a ha = units.mk0 b hb ↔ a = b :=
+⟨λ h, by injection h, λ h, units.ext h⟩
+
+end units
+
+namespace gwz
+variables {G : Type*} [group_with_zero G]
+
+lemma mul_eq_zero (a b : G) (h : a * b = 0) : a = 0 ∨ b = 0 :=
+begin
+  classical, contrapose! h,
+  exact unit_ne_zero ((units.mk0 a h.1) * (units.mk0 b h.2))
+end
+
+section prio
+set_option default_priority 10 -- see Note [default priority]
+
+instance : no_zero_divisors G :=
+{ eq_zero_or_eq_zero_of_mul_eq_zero := mul_eq_zero,
+  .. (‹_› : group_with_zero G) }
+
+end prio
+
+@[simp] lemma mul_eq_zero_iff {a b : G} : a * b = 0 ↔ a = 0 ∨ b = 0 :=
+⟨mul_eq_zero a b, by rintro (H|H); simp [H]⟩
+
+lemma mul_ne_zero {a b : G} (ha : a ≠ 0) (hb : b ≠ 0) : a * b ≠ 0 :=
+begin
+  assume ab0, rw gwz.mul_eq_zero_iff at ab0,
+  cases ab0; contradiction
+end
+
+lemma mul_ne_zero_iff {a b : G} : a * b ≠ 0 ↔ a ≠ 0 ∧ b ≠ 0 :=
+by { classical, rw ← not_iff_not, push_neg, exact mul_eq_zero_iff }
+
+lemma mul_left_cancel {x : G} (hx : x ≠ 0) {y z : G} (h : x * y = x * z) : y = z :=
+calc y = x⁻¹ * (x * y) : by rw inv_mul_cancel_assoc_right _ _ hx
+   ... = x⁻¹ * (x * z) : by rw h
+   ... = z             : by rw inv_mul_cancel_assoc_right _ _ hx
+
+lemma mul_right_cancel {x : G} (hx : x ≠ 0) {y z : G} (h : y * x = z * x) : y = z :=
+calc y = (y * x) * x⁻¹ : by rw mul_inv_cancel_assoc_left _ _ hx
+   ... = (z * x) * x⁻¹ : by rw h
+   ... = z             : by rw mul_inv_cancel_assoc_left _ _ hx
+
+lemma mul_inv_eq_of_eq_mul {x : G} (hx : x ≠ 0) {y z : G} (h : y = z * x) : y * x⁻¹ = z :=
+h.symm ▸ (mul_inv_cancel_assoc_left z x hx)
+
+lemma eq_mul_inv_of_mul_eq {x : G} (hx : x ≠ 0) {y z : G} (h : z * x = y) : z = y * x⁻¹ :=
+eq.symm $ mul_inv_eq_of_eq_mul hx h.symm
+
+lemma mul_inv_rev (x y : G) : (x * y)⁻¹ = y⁻¹ * x⁻¹ :=
+begin
+  classical,
+  by_cases hx : x = 0, { simp [hx] },
+  by_cases hy : y = 0, { simp [hy] },
+  symmetry,
+  apply eq_inv_of_mul_left_eq_one,
+  simp [mul_assoc, hx, hy]
+end
+
+theorem inv_comm_of_comm {a b : G} (H : a * b = b * a) : a⁻¹ * b = b * a⁻¹ :=
+begin
+  have : a⁻¹ * (b * a) * a⁻¹ = a⁻¹ * (a * b) * a⁻¹ :=
+    congr_arg (λ x:G, a⁻¹ * x * a⁻¹) H.symm,
+  classical,
+  by_cases h : a = 0, { rw [h, inv_zero, zero_mul, mul_zero] },
+  rwa [inv_mul_cancel_assoc_right _ _ h, mul_assoc, mul_inv_cancel_assoc_left _ _ h] at this,
+end
+
+@[simp] lemma div_self {a : G} (h : a ≠ 0) : a / a = 1 := mul_inv_cancel _ h
+
+@[simp] lemma div_one (a : G) : a / 1 = a :=
+show a * 1⁻¹ = a, by rw [inv_one, mul_one]
+
+lemma one_div (a : G) : 1 / a = a⁻¹ := one_mul _
```

Comment:
Which is the simp normal form? Is it useful to have 1/a's around? I would imagine that in the integers it's better to have `-x` then `0 - x`. Should `simp` be turning 1/a into a^{-1}?

Reply:
I've thought and asked about this before. It seems to be hard to get a satisfactory simp normal form here.

Reply:
Do you know what goes wrong with simplifying to a^{-1} here? It seems worth a try, but I don't know.